pkg/devicerepository/store/bleve/store.go

```diff
@@ -0,0 +1,192 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bleve
+
+import (
+	"github.com/blevesearch/bleve"
+	"github.com/blevesearch/bleve/search/query"
+	"go.thethings.network/lorawan-stack/v3/pkg/devicerepository/store"
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/jsonpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+var (
+	errCorruptedIndex = errors.DefineCorruption("corrupted_index", "corrupted index file")
+)
+
+// GetBrands lists available end device vendors.
+func (s *bleveStore) GetBrands(req store.GetBrandsRequest) (*store.GetBrandsResponse, error) {
+	queries := []query.Query{
+		bleve.NewMatchAllQuery(),
+	}
+	if q := req.Search; q != "" {
+		queries = append(queries, bleve.NewQueryStringQuery(q))
+	}
+	if q := req.BrandID; q != "" {
+		query := bleve.NewMatchQuery(q)
+		query.SetField("BrandID")
+		queries = append(queries, query)
+	}
+
+	searchRequest := bleve.NewSearchRequest(bleve.NewConjunctionQuery(queries...))
+	searchRequest.Size = int(req.Limit)
+	if req.Page == 0 {
+		req.Page = 1
+	}
+	searchRequest.From = int((req.Page - 1) * req.Limit)
+
+	searchRequest.Fields = []string{"BrandPB"}
+	switch req.OrderBy {
+	case "brand_id":
+		searchRequest.SortBy([]string{"BrandID"})
+	case "-brand_id":
+		searchRequest.SortBy([]string{"-BrandID"})
+	case "name":
+		searchRequest.SortBy([]string{"BrandName"})
+	case "-name":
+		searchRequest.SortBy([]string{"-BrandName"})
+	}
+
+	s.brandsIndexMu.RLock()
+	result, err := s.brandsIndex.Search(searchRequest)
+	if err != nil {
+		return nil, err
+	}
+	s.brandsIndexMu.RUnlock()
+
+	brands := make([]*ttnpb.EndDeviceBrand, 0, len(result.Hits))
+	for _, hit := range result.Hits {
+		s, ok := hit.Fields["BrandPB"].(string)
+		if !ok {
+			return nil, errCorruptedIndex.New()
+		}
+		brand := &ttnpb.EndDeviceBrand{}
+		if err := jsonpb.TTN().Unmarshal([]byte(s), brand); err != nil {
```

Comment:
Isn't it much cheaper to keep the original struct in memory? Now you're unmarshaling JSON for each hit, on every search request? With a fancy autocomplete mechanism, that could be dozens per second per user searching for a device.

Reply:
Quoting https://github.com/TheThingsNetwork/lorawan-stack/issues/3433#issuecomment-734128445:

> ... it is not trivial to load them back from the index. E.g. for an ttnpb.EndDevice object, we would have to retrieve every single nested field individually, and then recreate the original proto, which I would rather avoid....

This means that, instead of `BrandPB`, we would need to individually load all fields as an interface first, then manually set in the result proto. This seems very tedious to me.

For reference, here is what the fields for an end device model are (NOTE: These contain all key names, including `EU_863_870`, even though they are not part of the proto).

```
$ ./bleve check modelsIndex.bleve
checking fields: [ModelPB.brand_id ModelPB.model_id ModelPB.name ModelPB.description ModelPB.firmware_versions.version ModelPB.firmware_versions.numeric ModelPB.firmware_versions.profiles.EU_863_870.profile_id ModelPB.firmware_versions.profiles.EU_863_870.lorawan_certified ModelPB.firmware_versions.profiles.EU_863_870.codec_id ModelPB.ip_code ModelPB.key_security ModelPB.hardware_versions.version ModelPB.hardware_versions.numeric ModelPB.product_url ModelPB.datasheet_url BrandID ModelPB.hardware_versions.part_number BrandName ModelID ModelName _all ModelPB.firmware_versions.supported_hardware_versions ModelPB.firmware_versions.profiles.AS_923.profile_id ModelPB.firmware_versions.profiles.AS_923.lorawan_certified ModelPB.firmware_versions.profiles.AS_923.codec_id ModelPB.firmware_versions.profiles.US_902_928.profile_id  (... and lots more ....)]
....
```

Converting to and from JSON is not the most performant option, but improving on it is not trivial. I am OK with creating an issue to deal with this later, and adding a TODO comment instead of blocking on this.

Reply:
I don't get it really. You are searching through structs, right? There can be a pointer in there to an already unmarshaled entity, right?

Reply:
Well, it does not seem possible to store an actual struct (e.g. `*ttnpb.EndDeviceBrand`) and load it back from the index. We could instead add an in-memory cache for this, to avoid unmarshaling JSON repeatedly?